Enable NonEscapable types by default and move lifetime dependence support under a new feature

meg-gupta opened the pull request:
SE-0446 is accepted - Promote `NonescapableTypes` feature into a LANGUAGE_FEATURE. Move lifetime dependence support under a new experimental flag.

`LifetimeDependence` feature flag ensures Span's use of lifetime features are appropriately printed and work with older compilers which don't understand them yet.

meg-gupta committed:
Introduce LifetimeDependence experimental feature

meg-gupta committed:
Add -enable-experimental-feature LifetimeDependence to stdlib

meg-gupta committed:
Promote Nonescapable types to a language feature

meg-gupta committed:
Update tests

meg-gupta commented:
@swift-ci test

meg-gupta commented:
@swift-ci test

ahoppen commented:
Could you also enable the feature in swift-syntax by default?

tshortli commented on stdlib/cmake/modules/SwiftSource.cmake:
```diff
@@ -634,7 +634,7 @@ function(_compile_swift_files
   list(APPEND swift_flags "-enable-experimental-feature" "AllowUnsafeAttribute")
 
-  list(APPEND swift_flags "-enable-experimental-feature" "NonescapableTypes")
+  list(APPEND swift_flags "-enable-experimental-feature" "LifetimeDependence")
```
My expectation is that it is not safe to remove `NonescapableTypes` here until all older compilers we support have it enabled by default. Until then, `-enable-experimental-feature NonescapableTypes` needs to appear in the `.swiftinterface` so that the older compilers allow use of non-escapable types in the code in the interface.

atrick reviewed (commented):
No concerns from me!

meg-gupta committed:
Add an initializer for Span without lifetime dependence

Without this, supported old compilers which don't recognize LifetimeDependence feature but
recognize the NonEscapableTypes feature will run into lifetime dependence inference errors
on the implicit initializer generated for the stdlib's swiftinterface

meg-gupta commented:
@swift-ci test

meg-gupta commented:
preset=oss-swift-test-stdlib-with-toolchain
@swift-ci please test with toolchain and preset

meg-gupta commented:
@swift-ci test

meg-gupta commented:
@swift-ci test

meg-gupta commented:
@swift-ci test

meg-gupta commented:
@swift-ci test

meg-gupta commented:
@swift-ci test

Xazax-hun commented on stdlib/cmake/modules/SwiftSource.cmake:
```diff
@@ -634,6 +634,8 @@ function(_compile_swift_files
   list(APPEND swift_flags "-enable-experimental-feature" "SE427NoInferenceOnExtension")
   list(APPEND swift_flags "-enable-experimental-feature" "AllowUnsafeAttribute")
   list(APPEND swift_flags "-enable-experimental-feature" "NonescapableTypes")
```
Late to the party, but is there a reason why we kept this line?

tshortli replied:
See my resolved comment: https://github.com/swiftlang/swift/pull/77637#discussion_r1844274996

Xazax-hun replied:
Ah, I see, thanks!
